File: src/lib/chartSearch.ts
// Chart search client. Finds anything named in the NOAA ENC store — lights,
// wrecks, canyons, channels, anchorages, harbours — and gives back somewhere
// to look on the map.
//
// Search runs on the chart server (it owns the feature store); this module
// shapes the request, debounces the type-ahead, and works out how the map
// should frame a hit.

import { resolveTileBase } from "./autoRoute";

export interface SearchHit {
  name: string;
  /** S-57 object class acronym, e.g. "LIGHTS". */
  class: string;
  /** That class in words ("Light"), or the acronym when we have no wording. */
  label: string;
  cell: string;
  lat: number;
  lng: number;
  /** [minLon, minLat, maxLon, maxLat] — a feature's full extent. */
  bbox: [number, number, number, number];
  /** Metres from the origin passed to the search, or -1 when none was. */
  distance_meters: number;
}

/** Builds the /noaa-enc/search query. Exported for testing. */
export function searchUrl(
  base: string,
  q: string,
  opts: { origin?: { lat: number; lng: number } | null; limit?: number; objectClass?: string } = {}
): string {
  const p = new URLSearchParams({ q });
  if (opts.origin) {
    // Nearest-first only makes sense with somewhere to measure from. Without
    // it the server falls back to alphabetical.
    p.set("lat", String(opts.origin.lat));
    p.set("lon", String(opts.origin.lng));
  }
  if (opts.limit != null && opts.limit > 0) p.set("limit", String(opts.limit));
  if (opts.objectClass) p.set("class", opts.objectClass);
  return `${base}/noaa-enc/search?${p.toString()}`;
}

export interface SearchResponse {
  hits: SearchHit[];
  /**
   * The terms that actually matched. When it differs from what was typed, the
   * full phrase found nothing and this is a narrower answer — say so rather
   * than present it as an exact match.
   */
  matchedQuery: string;
}

export async function searchChart(
  q: string,
  opts: { origin?: { lat: number; lng: number } | null; limit?: number; objectClass?: string } = {}
): Promise<SearchResponse> {
  const trimmed = q.trim();
  if (!trimmed) return { hits: [], matchedQuery: trimmed };
  const base = await resolveTileBase();
  const resp = await fetch(searchUrl(base, trimmed, opts));
  if (!resp.ok) {
    let msg = `search failed (${resp.status})`;
    try {
      const body = await resp.json();
      if (body && typeof body.error === "string") msg = body.error;
    } catch {
      // non-JSON error body; keep the status message
    }
    throw new Error(msg);
  }
  const body = await resp.json();
  return {
    hits: (body?.results ?? []) as SearchHit[],
    matchedQuery: typeof body?.matched_query === "string" ? body.matched_query : trimmed,
  };
}

/** Shortest query worth sending. One or two letters match half the chart. */
export const MIN_QUERY_LENGTH = 3;

/**
 * How the map should frame a hit. A point feature (a buoy, a light) has no
 * extent, so centre on it at a close zoom; an area (a canyon, a channel) does,
 * so fit the whole thing.
 */
export function framingFor(hit: SearchHit): { kind: "center"; zoom: number } | { kind: "fit" } {
  const [minLon, minLat, maxLon, maxLat] = hit.bbox;
  const spanLon = Math.abs(maxLon - minLon);
  const spanLat = Math.abs(maxLat - minLat);
  // Under ~50 m across in either axis there is nothing to fit to — fitting a
  // degenerate extent zooms to maximum and shows a blank tile.
  const DEGENERATE_DEG = 0.0005;
  if (spanLon < DEGENERATE_DEG && spanLat < DEGENERATE_DEG) {
    return { kind: "center", zoom: 15 };
  }
  return { kind: "fit" };
}

/** Formats a hit's distance for the result row; empty when unknown. */
export function formatDistance(meters: number): string {
  if (!(meters >= 0)) return "";
  const nm = meters / 1852;
  return nm < 10 ? `${nm.toFixed(1)} nm` : `${Math.round(nm)} nm`;
}

/**
 * Debounces an async search so typing doesn't fire a request per keystroke,
 * and so a slow earlier response can never overwrite a newer one.
 */
export function makeSearchRunner<T>(
  run: (q: string) => Promise<T>,
  /** What to report for a query too short to send. */
  empty: T,
  delayMs = 250
): {
  search: (
    q: string,
    onResult: (result: T, q: string) => void,
    onError: (e: Error) => void
  ) => void;
  cancel: () => void;
} {
  let timer: ReturnType<typeof setTimeout> | undefined;
  let seq = 0;

  return {
    search(q, onResult, onError) {
      if (timer) clearTimeout(timer);
      const mine = ++seq; // stamp: a stale response is dropped, not rendered
      if (q.trim().length < MIN_QUERY_LENGTH) {
        onResult(empty, q);
        return;
      }
      timer = setTimeout(() => {
        run(q).then(
          (result) => {
            if (mine === seq) onResult(result, q);
          },
          (e) => {
            if (mine === seq) onError(e instanceof Error ? e : new Error(String(e)));
          }
        );
      }, delayMs);
    },
    cancel() {
      if (timer) clearTimeout(timer);
      seq++; // invalidate anything in flight
    },
  };
}

File: mobile/lib/chart/chart_search.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:latlong2/latlong.dart';

/// Chart search client, ported from src/lib/chartSearch.ts (tests translated
/// alongside). Finds anything named in the NOAA ENC store — lights, wrecks,
/// canyons, channels, anchorages, harbours — and gives back somewhere to look
/// on the map. Search runs on the chart server (it owns the feature store);
/// this module shapes the request, debounces the type-ahead, and works out
/// how the map should frame a hit.

class SearchHit {
  const SearchHit({
    required this.name,
    required this.class_,
    required this.label,
    required this.cell,
    required this.lat,
    required this.lng,
    required this.bbox,
    required this.distanceMeters,
    this.area = '',
  });

  final String name;

  /// S-57 object class acronym, e.g. "LIGHTS".
  final String class_;

  /// That class in words ("Light"), or the acronym when we have no wording.
  final String label;
  final String cell;
  final double lat;
  final double lng;

  /// [minLon, minLat, maxLon, maxLat] — the feature's full extent.
  final List<double> bbox;

  /// Metres from the origin passed to the search, or -1 when none was.
  final double distanceMeters;

  /// Where the hit is for a human — "Newport, RI" — or empty when the
  /// server could not place it.
  final String area;

  LatLng get pos => LatLng(lat, lng);

  static SearchHit? fromJson(dynamic j) {
    if (j is! Map) return null;
    final lat = j['lat'];
    final lng = j['lng'];
    if (lat is! num || lng is! num) return null;
    final rawBbox = j['bbox'];
    final bbox = [
      if (rawBbox is List)
        for (final v in rawBbox)
          if (v is num) v.toDouble()
    ];
    return SearchHit(
      name: (j['name'] ?? '').toString(),
      class_: (j['class'] ?? '').toString(),
      label: (j['label'] ?? j['class'] ?? '').toString(),
      cell: (j['cell'] ?? '').toString(),
      lat: lat.toDouble(),
      lng: lng.toDouble(),
      bbox: bbox.length == 4
          ? bbox
          : [lng.toDouble(), lat.toDouble(), lng.toDouble(), lat.toDouble()],
      distanceMeters: (j['distance_meters'] as num?)?.toDouble() ?? -1,
      area: (j['area'] ?? '').toString(),
    );
  }
}

/// Builds the /noaa-enc/search query. Exposed for testing.
String searchUrl(
  String base,
  String q, {
  LatLng? origin,
  int? limit,
  String? objectClass,
}) {
  final p = <String, String>{'q': q};
  if (origin != null) {
    // Nearest-first only makes sense with somewhere to measure from. Without
    // it the server falls back to alphabetical.
    p['lat'] = '${origin.latitude}';
    p['lon'] = '${origin.longitude}';
  }
  if (limit != null && limit > 0) p['limit'] = '$limit';
  if (objectClass != null && objectClass.isNotEmpty) p['class'] = objectClass;
  return '$base/noaa-enc/search?${Uri(queryParameters: p).query}';
}

class SearchResponse {
  const SearchResponse({required this.hits, required this.matchedQuery});
  final List<SearchHit> hits;

  /// The terms that actually matched. When it differs from what was typed,
  /// the full phrase found nothing and this is a narrower answer — say so
  /// rather than present it as an exact match.
  final String matchedQuery;
}

Future<SearchResponse> searchChart(
  String base,
  String q, {
  LatLng? origin,
  int? limit,
  String? objectClass,
  http.Client? client,
}) async {
  final trimmed = q.trim();
  if (trimmed.isEmpty) return const SearchResponse(hits: [], matchedQuery: '');
  final url = Uri.parse(
      searchUrl(base, trimmed, origin: origin, limit: limit, objectClass: objectClass));
  final resp = await (client?.get(url) ?? http.get(url));
  if (resp.statusCode != 200) {
    var msg = 'search failed (${resp.statusCode})';
    try {
      final body = jsonDecode(resp.body);
      if (body is Map && body['error'] is String) msg = body['error'] as String;
    } catch (_) {
      // non-JSON error body; keep the status message
    }
    throw Exception(msg);
  }
  final body = jsonDecode(resp.body);
  final results = body is Map ? body['results'] : null;
  return SearchResponse(
    hits: [
      if (results is List)
        for (final r in results)
          if (SearchHit.fromJson(r) case final hit?) hit
    ],
    matchedQuery: body is Map && body['matched_query'] is String
        ? body['matched_query'] as String
        : trimmed,
  );
}

/// Shortest query worth sending. One or two letters match half the chart.
const int minQueryLength = 3;

/// How the map should frame a hit: a point feature (a buoy, a light) has no
/// extent, so centre on it at a close zoom; an area (a canyon, a channel)
/// does, so fit the whole thing. Under ~50 m across in either axis there is
/// nothing to fit to — fitting a degenerate extent zooms to maximum and
/// shows a blank tile.
({bool fit, double zoom}) framingFor(SearchHit hit) {
  const degenerateDeg = 0.0005;
  final spanLon = (hit.bbox[2] - hit.bbox[0]).abs();
  final spanLat = (hit.bbox[3] - hit.bbox[1]).abs();
  if (spanLon < degenerateDeg && spanLat < degenerateDeg) {
    return (fit: false, zoom: 15);
  }
  return (fit: true, zoom: 0);
}

/// Formats a hit's distance for the result row; empty when unknown.
String formatSearchDistance(double meters) {
  if (!(meters >= 0)) return '';
  final nm = meters / 1852;
  return nm < 10 ? '${nm.toStringAsFixed(1)} nm' : '${nm.round()} nm';
}

/// Debounces an async search so typing doesn't fire a request per keystroke,
/// and so a slow earlier response can never overwrite a newer one.
class SearchRunner<T> {
  SearchRunner(this._run, this._empty,
      {this.delay = const Duration(milliseconds: 250)});

  final Future<T> Function(String q) _run;

  /// What to report for a query too short to send.
  final T _empty;
  final Duration delay;

  Timer? _timer;
  int _seq = 0;

  void search(
    String q,
    void Function(T result, String q) onResult,
    void Function(Object e) onError,
  ) {
    _timer?.cancel();
    final mine = ++_seq; // stamp: a stale response is dropped, not rendered
    if (q.trim().length < minQueryLength) {
      onResult(_empty, q);
      return;
    }
    _timer = Timer(delay, () {
      _run(q).then(
        (result) {
          if (mine == _seq) onResult(result, q);
        },
        onError: (Object e) {
          if (mine == _seq) onError(e);
        },
      );
    });
  }

  void cancel() {
    _timer?.cancel();
    _seq++; // invalidate anything in flight
  }
}
